plugins/auto-sizes/hooks.php

```diff
@@ -64,28 +64,46 @@ function auto_sizes_update_content_img_tag( $html ): string {
 		return $html;
 	}
 
-	// Bail early if the image is not lazy-loaded.
-	$value = $processor->get_attribute( 'loading' );
-	if ( ! is_string( $value ) || 'lazy' !== strtolower( trim( $value, " \t\f\r\n" ) ) ) {
-		return $html;
+	auto_sizes_apply_tag_parser_updates( $processor );
+	return $processor->get_updated_html();
+}
+add_filter( 'wp_content_img_tag', 'auto_sizes_update_content_img_tag' );
+
+/**
+ * Applies changes to current IMG tag to add or remove 'auto' from the sizes attribute when needed.
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @param WP_HTML_Tag_Processor $processor HTML processor currently at an IMG tag.
+ * @return bool Whether changes were made.
+ */
+function auto_sizes_apply_tag_parser_updates( WP_HTML_Tag_Processor $processor ): bool {
+	if ( $processor->get_tag() !== 'IMG' ) {
+		return false;
 	}
 
 	$sizes = $processor->get_attribute( 'sizes' );
-
-	// Bail early if the image is not responsive.
 	if ( ! is_string( $sizes ) ) {
-		return $html;
+		return false;
 	}
 
-	// Don't add 'auto' to the sizes attribute if it already exists.
-	if ( auto_sizes_attribute_includes_valid_auto( $sizes ) ) {
-		return $html;
-	}
+	$value   = $processor->get_attribute( 'loading' );
+	$is_lazy = ( is_string( $value ) && 'lazy' === strtolower( trim( $value, " \t\f\r\n" ) ) );
 
-	$processor->set_attribute( 'sizes', "auto, $sizes" );
-	return $processor->get_updated_html();
+	$has_auto = auto_sizes_attribute_includes_valid_auto( $sizes );
+
+	if ( $is_lazy && ! $has_auto ) {
+		$processor->set_attribute( 'sizes', "auto, $sizes" );
+	} elseif ( ! $is_lazy && $has_auto ) {
+		// Remove auto from the beginning of the list.
+		$processor->set_attribute(
+			'sizes',
+			(string) preg_replace( '/^[ \t\f\r\n]*auto[ \t\f\r\n]*(,[ \t\f\r\n]*)?/i', '', $sizes )
+		);
```

Comment:
at this point it seems like there's some zig-zagging going on for the `auto` keyword. just going to toss this out there, but we're using one method to detect its presence, and another method to add it, but I wonder if the PCRE method doesn't already in some way capture both situations, and might not be clearer to rely on vs. splitting the detection and replacement.

```php
function toggle_auto_on_sizes_attribute( string $existing, bool $wants_auto ): string {
	// Find any existing `auto` keyword at the front of the attribute.
	preg_match( '/^([ \t\f\r\n]*auto[ \t\r\f\n]*)(?:,|$)/i', $existing, $auto_match, PREG_OFFSET_CAPTURE );
	$has_auto = isset( $auto_match[1][0] );

	if ( $wants_auto && ! $has_auto ) {
		return "auto, {$sizes}";
	} elseif ( ! $wants_auto && $has_auto ) {
		// Chop off leading `auto` keyword.
		return substr( $existing, strlen( $auto_match[1][0] ) );
	}

	return $existing;
}
```

Then the calling code uses `$is_lazy` as a proxy for `$wants_auto`.

```php
$processor->set_attribute( 'sizes', toggle_auto_on_sizes_attribute( $sizes, $is_lazy ) );
```